### PR TITLE
LPS-116201 Update liferay-ckeditor version in portal and implement changes to make it fit with new skin

### PR DESCRIPTION
**Pull request by carloslancha**
https://issues.liferay.com/browse/LPS-116201

Updating `liferay-ckeditor` version in portal in order to apply the new skin we created following Lexicon's specs.

icon states

Test plan:

- Go to Product Menu / Content & Data / Web Content / New Basic Web Content (or ay other place with an editor)
- Play with the editor, check the icons and the styling.

### CODE REVIEW

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
@jbalsas should we wait until GA4 is released before mergin this?

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by wincent** [comment]
As per [Slack thread](https://liferay.slack.com/archives/G3JBR21HA/p1593505265243300?thread_ts=1593465518.238800&cid=G3JBR21HA), we might want to merge https://github.com/liferay/liferay-ckeditor/pull/78 and cut a release so as to avoid adding to the dependency foot-print. (At least, it's my understanding that the new deps added to liferay-ckeditor should be build-only deps.)

**Comment by wincent** [comment]
> should we wait until GA4 is released before merging this?

Currently [targeting Thursday for release](https://liferay.slack.com/archives/C03BTCQBQ/p1593466292125600), so I'd say it is quite reasonable to wait a couple of days if you feel like there is some risk involved here. FWIW, if you've manually tested it in a few places *and* checked IE, then it is probably ok.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java [hunk 1]
```diff
@@ -60,7 +60,8 @@ public void populateConfigJSONObject(
 				HtmlUtil.escape(
 					PortalUtil.getStaticResourceURL(
 						themeDisplay.getRequest(),
-						"/o/frontend-css-web/main.css")))
+						"/o/frontend-editor-ckeditor-web/ckeditor/skins" +
```
Comment (jbalsas) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:63]: Is this really needed? I'd expect CKEditor to load this automatically when passing in a different skin?

Comment (carloslancha) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:63]: Checked, we don't need this :D

Comment (jbalsas) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:63]: Did we set the skin to be the default in our build process in `liferay-ckeditor`?

File: modules/yarn.lock [hunk 2]
```diff
@@ -5324,6 +5333,11 @@ chownr@^1.0.1, chownr@^1.1.1, chownr@^1.1.2:
   resolved "https://registry.yarnpkg.com/chownr/-/chownr-1.1.3.tgz#42d837d5239688d55f303003a508230fa6727142"
   integrity sha512-i70fVHhmV3DtTl6nqvZOnIjbY0Pe4kAUjwHj8z0zAdgBtYrJyYwLKCCuRBQ5ppkyL0AkN7HKRnETdmdp1zqNXw==
 
+chownr@^2.0.0:
```
Comment (jbalsas) [inline modules/yarn.lock:5336]: 👀

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java (hunk 1 above)
Comment (carloslancha) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:63]: Yep https://github.com/liferay/liferay-ckeditor/blob/master/support/build-config.js#L9

**Comment by carloslancha** [comment]
@wincent @jbalsas released `liferay-ckeditor.4.13.1-liferay.8` and updated it in portal

**Comment by carloslancha** [comment]
ci:test:sf

File: modules/yarn.lock (hunk 2 above)
Comment (carloslancha) [reply modules/yarn.lock]: Removed in last commit

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by wincent** [comment]
I'll look at then code in a bit, but can you please retitle the PR so that it isn't just "SF"? Thanks.

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by wincent** [comment]
Lockfile update looks better this time!

I noticed in [the commits list of the liferay-ckeditor repo](https://github.com/liferay/liferay-ckeditor/commits/master) that we did a build before the release, and then a revert and rebuild after the release:

```
f3273d3b70 chore: build ckeditor (master) (16 hours ago)
28204e2575 Revert "chore: build ckeditor" (16 hours ago)
2108b86398 chore: prepare v4.13.1-liferay.8 release (tag: v4.13.1-liferay.8) (16 hours ago)
faf571b2ac chore: build ckeditor (16 hours ago)
```

Is it just the unwanted submodule update that was bad and had to be reverted, and we don't need to cut another release, right?

```
$ git diff --ignore-submodules=none 2108b86398..f3273d3b70
diff --git a/ckeditor-dev b/ckeditor-dev
index 24e0ea7a5a..411985373e 160000
--- a/ckeditor-dev
+++ b/ckeditor-dev
@@ -1 +1 @@
-Subproject commit 24e0ea7a5a41093c605bfdbc59b6d792143972bc
+Subproject commit 411985373e833a0df76e4ebd0afb45f337612725
```

What was the cause of the unwanted update? We want to "foolproof" the build script and the instructions in the README as much as possible to make it harder to make mistakes.

**Comment by wincent** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
@wincent exactly, `ckeditor-dev` submodule changes were introduced in that commit (which I didn't pretend) and caused ci failures. Since that submodule doesn't interfere with the release itself I just hotfixed it.

I usually make the commits from VSCode and maybe the extensión doesn't work well with submodules and added it without showing the folder in the list. I was thinking on adding a note in the README.md just to be careful with that.

**Comment by wincent** [comment]
> I usually make the commits from VSCode and maybe the extensión doesn't work well with submodules and added it without showing the folder in the list. I was thinking on adding a note in the README.md just to be careful with that.

Not a bad idea. We have a long history of mistakes with that submodule that we have tried to fix by moving more and more logic into the `ck.sh` script itself; for example, current state is [this](https://github.com/liferay/liferay-ckeditor/blob/6d6a66828157508a7ccf91ef5711acf88526ea6c/ck.sh#L81-L90):

```
git add ckeditor

echo
echo "✅ DONE"
echo
echo "Updated build artifacts have been staged, ready to commit."
echo
echo "You can review them with \`git status\` and \`git diff --cached\`"
echo
```

Note the `git add ckeditor`; we do everything but actually create the commit for you — maybe we should do that too... Thoughts? cc @julien

**Comment by julien** [comment]
@wincent, @carloslancha: It's true that we've made a lot of mistakes with the submodule, so if we can do something to avoid it, I'm all for it.

**Comment by carloslancha** [comment]
I don't think creating the commit is a good idea, during the skin development I ran `sh ck.sh` every 5 minutes 😅 and I didn't want to commit every result

**Comment by wincent** [comment]
> I don't think creating the commit is a good idea

Yeah, that's why we didn't do it so far. But it's a trade-off. Would you rather have:

- No submodule corruption, but a lot of autocommits that you can squash before sending?
- No autocommits, but periodically corrupt the submodule?

I was hoping we could make the second one work (by learning not to screw up the submodule over time), but I am wondering whether that was optimistic. If it's VSCode that's producing the bad commits (because it is effectively doing an `git commit -a` and not showing what's going to happen in the UI), then it's not even a question of user error: you can't use education to fix a ~bug~ problem in VSCode.

**Comment by carloslancha** [comment]
What about creating a publish script?

```
publish)
		echo
		echo "⚠️  WARNING"
		echo

		echo "This will clean all your uncommited files"
		read -r -p "Are you sure you want to continue? [y/n] " yn

		case $yn in
			[Yy]*)
				git checkout master
				git pull upstream master --ff-only
				git status

				sh ck.sh build

				yarn ci

				git commit -m "chore: build CKEditor"

				read -r -p "Choose a version number " VERSION

				npx liferay-changelog-generator --version=$VERSION

				yarn version --new-version $VERSION
				;;
			*)
				echo
				echo "Aborting."
				echo
				exit
		esac
		;;
```

**Comment by wincent** [comment]
> What about creating a publish script?

You'd need a `git add` in there too to include the changelog, but *that's* exactly the kind of step that I'd want to be manual; ie. once you run `yarn version` there is no going back (without yanking), so you really do want the user to be the one running `yarn version` so that they can verify that all previous steps have been performed — and performed with the expected results —before it happens.

**Comment by carloslancha** [comment]
If we want that to be manual then I'm ok with having a lot of "build ckeditor" commits you will need to squash or remove, moreover during development those commits will pass through a pr review process so it's more difficult that could make it to master.

**Comment by wincent** [comment]
> If we want that to be manual then I'm ok with having a lot of "build ckeditor" commits you will need to squash or remove, moreover during development those commits will pass through a pr review process so it's more difficult that could make it to master.

Ok, let's give it a try. I'll send a PR. We can always change our mind later if it proves to be a terrible idea.

**Comment by wincent** [comment]
Asked about unresponsive CI [here](https://liferay.slack.com/archives/C19PWQ74J/p1593617024078800).

**Comment by wincent** [comment]
From `#ci`:

> Looks like the `ci:test:sf` didn't get picked up by the GitHub webhook (possibly a GitHub issue). Try triggering it again.

**Comment by wincent** [comment]
ci:test:sf

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java [hunk 3]
```diff
@@ -80,11 +76,7 @@ public void populateConfigJSONObject(
 					PortalUtil.getStaticResourceURL(
 						themeDisplay.getRequest(),
 						"/o/frontend-editor-ckeditor-web/ckeditor/skins" +
-							"/moono-lisa/editor.css")),
-				HtmlUtil.escape(
-					PortalUtil.getStaticResourceURL(
-						themeDisplay.getRequest(),
-						"/o/frontend-css-web/main.css")))
+							"/moono-lexicon/editor.css")))
```
Comment (jbalsas) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: Interesting that we need this here... 🤔

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Hey @carloslancha, the test errors here seem related to the fix... can you take a look?

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java (hunk 3 above)
Comment (jbalsas) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: Do we need this here?

Comment (carloslancha) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: Let me check again

Comment (carloslancha) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: Yep, we need it:

image

Comment (jbalsas) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: So how did it use to work before? Were we also adding the `moono.css` file manually? It doesn't feel 100% right. I'd expect CKEditor's `ContextMenu` plugin to be already fetching the skin... but I could be wrong here...

Comment (carloslancha) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/BaseCKEditorConfigContributor.java:79]: Oh, wait, before applying the new styles we were not passing any `contextmenu_contentsCss` param to CKEditor, we added it to apply the styles from portal, probably we're overwriting it. Checking what happen if we remove the param.

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
Hey @john-co could you please take a look at the failing tests? What we're changing here is CKEditor but failing tests are in blogs, that is the only place where CKEditor is not being used but AlloyEditor. I tried to reproduce the failures locally with no success. Thx!

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
The test failure only happens on test runs from this PR based on testray history...but you're right that Blogs still uses Alloy Editor
image

@zhangyan0701 can you please investigate further?

**Comment by jbalsas** [comment]
Hey @carloslancha, @john-co, keep in mind that `AlloyEditor` is UI on top of CKEditor and as such, will use the CKEditor version we provide. Updates to CKEditor can thus produce failures in AlloyEditor tests.

Also, @julien tried updating to a newer version at https://github.com/liferay-frontend/liferay-portal/pull/86. Let's push it here so we can see if the issue is still present or try to figure it out otherwise!

**Comment by zhangyan0701** [comment]
There are some issues when I manually adding Blogs Entry locally with the same bundle as in this PR:

- The Title field and Subtitle field are not editable.
- The Content field will not extend if type into wrapping text.
- All buttons at the bottom of page are unavailable.
add blogs

See the second and the third issue in Web Content as well: Content field cannot extend. The Publish and Save as Draft buttons don't work.
add web content

**Comment by julien** [comment]
@zhangyan0701 thanks for reporting this.

In blogs, I'm also seeing that the *Title* and *Subtitle* fields aren't editable, I do have AlloyEditor's buttons enabled in the *Content* field but the content field is very small and the text seems to disappear when using the enter key (same issue in Web Content)

**Comment by zhangyan0701** [comment]
> the text seems to disappear when using the enter key (same issue in Web Content)

@julien Yes, if the text are with multiple lines, in content field only display the last line, other lines are hidden behind. I can see them if click upward arrow.

**Comment by julien** [comment]
I reset back to [`1ab1f36751d309929f4dab790558b24b8e124f3f`](https://github.com/liferay-frontend/liferay-portal/commit/1ab1f36751d309929f4dab790558b24b8e124f3f) did an `ant all` and `blogs-web` and `journal-web` are working as expected.

I'm also checked each commit and deployed to see what could have broken this, here's what I see:

- If I reset back to [`39ffe680748e67d584f4f215df7b4acb4267b312`](https://github.com/liferay-frontend/liferay-portal/commit/39ffe680748e67d584f4f215df7b4acb4267b312)  and deploy the changes, I don't see any issue

- If I reset back to [`fe7e7070257d7ca4de98c0da725215abac87343d`](https://github.com/liferay-frontend/liferay-portal/commit/fe7e7070257d7ca4de98c0da725215abac87343d),  and deploy the changes, there are no major issues, but in `journal-web` we do get this error although it doesn't prevent the application from working

```
Refused to apply style from 'http://localhost:8080/o/frontend-editor-ckeditor-web/ckeditor/skins/moono-lexicon/editor.css?browserId=chrome&themeId=admin_WAR_admintheme&languageId=en_US&b=7310&t=1594203674000' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
```

- If I reset back to [` 6679031542b3c4a29b8e94822d8b5c6bc069e296`](https://github.com/liferay-frontend/liferay-portal/commit/6679031542b3c4a29b8e94822d8b5c6bc069e296), and deploy the changes, that's when it seems to break 

blogs-web-error-0

`journal-web` also looks a bit broken, but works

web-content-error-0


 I'm going to revert [that](https://github.com/liferay-frontend/liferay-portal/commit/6679031542b3c4a29b8e94822d8b5c6bc069e296) commit, and hope it fixes something

**Comment by julien** [comment]
Reverted [`6679031542b3c4a29b8e94822d8b5c6bc069e296`](https://github.com/liferay-frontend/liferay-portal/commit/6679031542b3c4a29b8e94822d8b5c6bc069e296), did an `ant all` (delete and re-created the database just in case), and it's still broken.

Reverted [`fe7e7070257d7ca4de98c0da725215abac87343d`](https://github.com/liferay-frontend/liferay-portal/pull/66/commits/fe7e7070257d7ca4de98c0da725215abac87343d): still broken :cry: 

Reverted [`fd3105c3637097d78d6f9ca64c9c3612a00fe31b`](https://github.com/liferay-frontend/liferay-portal/pull/66/commits/fd3105c3637097d78d6f9ca64c9c3612a00fe31b): still broken :crying_cat_face: 

So up to now the only way I've managed to get if fixed is by using  `liferay-ckeditor@4.13.1-liferay.5` (currently used in `master`) and/or `liferay-ckeditor@4.13.1-liferay.6`: these two versions both use CKEDITOR 4.13.1 DEV - which is the also what `liferay-ckeditor@4.13.1-liferay.7` uses: I thought it might have been the update to CKEDITOR 4.14.1 done for [`liferay-ckeditor@4.14.1-liferay.1`](https://github.com/liferay/liferay-ckeditor/releases/tag/v4.14.1-liferay.1) that broke everything but it's isn't :sweat: 


So I guess now we need to look into `liferay-ckeditor` to see what could have happened

**Comment by julien** [comment]
I'm still looking at this, just wanted to add that I also tried reverting to `alloyeditor@2.11.8` (because that's what used in blogs and web content) so I just wanted to see if that made a difference but unfortunately it doesn't solve the issue 

You'll have to trust me
 
screenshot

On the `liferay-ckeditor` side, I've been able to check that with [52794df20](https://github.com/liferay/liferay-ckeditor/commit/52794df20) we still didn't have any issues, but I'll keep on looking

**Comment by wincent** [comment]
> I'm still looking at this, just wanted to add that I also tried reverting to `alloyeditor@2.11.8` (because that's what used in blogs and web content) so I just wanted to see if that made a difference but unfortunately it doesn't solve the issue

Well that is useful information, isn't it? Because the AlloyEditor build that we use in liferay-portal is the ["without CKEditor" variant](https://github.com/liferay/liferay-portal/blob/3994ec2ec702175c32939899392b6159acc423f3/modules/apps/frontend-editor/frontend-editor-alloyeditor-web/src/main/resources/META-INF/resources/resources.jsp#L39), which means that the problem is very likely local to our CKEditor build.

**Comment by julien** [comment]
@wincent yes I forgot to mention that is was the  ["without CKEditor" variant](https://github.com/liferay/liferay-portal/blob/3994ec2ec702175c32939899392b6159acc423f3/modules/apps/frontend-editor/frontend-editor-alloyeditor-web/src/main/resources/META-INF/resources/resources.jsp#L39), (which is the useful part) ... thanks for confirming too.

**Comment by jbalsas** [comment]
I get the feeling this is mostly a CSS issue... maybe with some sparkles of something else.

I still see CSS applying to `.alloy-editor` coming from `frontend-css-web`.

There's also a specific thing that looks odd in the `.cke_editable` css:

```css
.cke_editable {
    background-color: #f1f2f5;
    height: calc(100% - 12px)!important;
    margin: 0;
    padding: 12px 16px 8px 16px;
    position: absolute;
    width: calc(100% - 12px)!important;
}
```

Both the `width` and `height`, with `!important` and the `position: absolute` are a bit smelly... will need to compare with another branch, but something smells fishy here.

**Comment by jbalsas** [comment]
FYI, another thing that doesn't work is clicking with the mouse in the editable area. You can't get a cursor in the editable area with the mouse. Only if you use the arrow keys.

**Comment by jbalsas** [comment]
So, just to confirm, the old `moono-lisa` doesn't seem to have that `.cke_editable` portion, so it won't affect AlloyEditor, but the new `moono-lexicon` does... 🤷

**Comment by carloslancha** [comment]
I think I found the issue, I need to do a couple of tests but I have to leave now. I'll test it later and leave here what I found.

**Comment by carloslancha** [comment]
@jbalsas yep, it's because the skin, I didn't take into account that alloy editor uses portal's ckeditor version... 🤦

**Comment by julien** [comment]
Thanks @carloslancha and @jbalsas

**Comment by jbalsas** [comment]
> Use moono-lisa skin

monalisa

**Comment by carloslancha** [comment]
Yep, makes more sense using moono-lisa in alloy editor. Thx @julien !

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by julien** [comment]
@carloslancha no problem - I should have thought about this yesterday instead of checking all those commits :rofl:  let's see if this also fixes the CI issues :pray:

**Comment by carloslancha** [comment]
> no problem - I should have thought about this yesterday instead of checking all those commits 🤣 let's see if this also fixes the CI issues

I didn't think about that either and spent hours testing different commits so... 🤷 

You found it, that's what matters!

image

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by julien** [comment]
I'm not sure the last CI error is related to our change, but if @john-co can confirm that would be great

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Running it again, just in case... @john-co, could you please do some manual smoke tests on this, just to be sure?

**Comment by john-co** [comment]
It looks like the same error is coming from tests ran on test-1-17 similar to https://github.com/liferay-frontend/liferay-portal/pull/90#issuecomment-656168551. test-1-17 is now taken offline, and I think it shouldn't affect the current pending run.

**Comment by jbalsas** [comment]
ci:forward

**Comment by carloslancha** [comment]
Thx for the review @john-co and @jbalsas for the nightly forward!

**Pull request closed without merging**